psydac/linalg/topetsc.py

```diff
@@ -0,0 +1,81 @@
+import numpy as np
+
+from psydac.linalg.block import BlockVectorSpace, BlockVector, BlockMatrix
+from psydac.linalg.stencil import StencilVectorSpace, StencilVector
+from scipy.sparse import coo_matrix, bmat
+
+from mpi4py import MPI
+
+def get_npts( space ):
+
+    if isinstance(space, StencilVectorSpace):
+        npts = np.product(space.npts)
+    elif isinstance(space, BlockVectorSpace):
+        npts = sum([get_npts(s) for s in space.spaces])
+    else:
+        raise TypeError("Expected StencilVectorSpace or BlockVectorSpace, found instead {}".format(type(space)))
+
+    return npts
+
+def vec_tocoo( vec ):
+
+    if isinstance(vec, StencilVector):
+        npts = vec.space.npts
+        idx = tuple( slice(m*p,-m*p) for m,p in zip(vec.pads, vec.space.shifts) )
+        shape = vec._data[idx].shape
+        starts = vec.space.starts
+        indices = np.array([np.ravel_multi_index( [s+x for s,x in zip(starts, xx)], dims=npts,  order='C' ) for xx in np.ndindex(*shape)] )
+        data = vec._data[idx].flatten()
+        vec = coo_matrix(
+                    (data,(indices,indices)),
+                    shape = [np.prod(npts),np.prod(npts)],
+                    dtype = vec.space.dtype)
+    elif isinstance(vec, BlockVector):
+        vecs = [vec_tocoo(b) for b in vec.blocks]
+        blocks = [[None]*len(vecs) for v in vecs]
+        for i,v in enumerate(vecs):
+            blocks[i][i] = v
+        return bmat(blocks,format='coo')
+    else:
+        raise TypeError("Expected StencilVector or BlockVector, found instead {}".format(type(vec)))
+    return vec
+
+def vec_topetsc( vec ):
+    """ Convert to petsc data structure.
+    """
+
+    from petsc4py import PETSc
+    comm = vec.space.spaces[0].cart.global_comm if isinstance(vec, BlockVector) else vec.space.cart.global_comm
+    globalsize = get_npts(vec.space)
+    vec = vec_tocoo(vec)
+    gvec  = PETSc.Vec().create(comm=comm)
+    gvec.setSizes(globalsize)
+    gvec.setFromOptions()
+    gvec.setValues(vec.row, vec.data)
+    gvec.assemble()
+    return gvec
+
+def mat_topetsc( mat ):
+    """ Convert to petsc data structure.
+    """
```

Comment:
Please extend docstring

Reply:
done in [68937d4](https://github.com/pyccel/psydac/pull/244/commits/68937d4ffbcceec954f6f2e2bef2287f88f62853)